# 1.0 Business Understanding

### 1.1 Background

What is customer churn? This is the rate at which customers leave a business(especially SaaS)<br>
SyriaTel is a telecommunications company facing the challenge of customer churn, which refers to customers discontinuing their services. Churn can have significant financial implications for SyriaTel, including the loss of recurring revenue, increased customer acquisition costs, and potential negative impact on the company's reputation. To address this issue, SyriaTel aims to build a churn prediction system that can identify customers likely to churn in the near future.

According to Forbes Advisor an article by Monique Danao, published March 2nd 2023,11.00am, Customer churn rate, "refers to the rate at which subscribers or customers stop transacting with your business." https://www.forbes.com/advisor/business/churn-rate/

In other words customers discontinue the use of company services and products which in turn leads to revenue loaas and in the long run affects the companys profitability.


## Project Questions
* Are there any predictable patterns to whether a customer will soon stop doing business with SyriaTel

## Objectives
* To determine whether customers who have a high number of outbound calls or data usage are likely to churn
* To identify if the number of customer service calls has a relationship to churn
* To asses whether the number of outbound calls or data usage affects churn
* To determine if the time of day influences churn

## Stakeholders
SyriaTel Telecommunications Company
Focusing on the 3 departments :
 * Top Management
 * Marketing Team
 * Customer Retention Team

## Hypothesis
H0: The number of customer service calls has no relationship to churn<br>
H1: The number of customer service calls is related to churn

Customers who have a shorter tenure with SyriaTel are more likely to churn <br>
Customers who have a high number of outbound calls or data usage are less likely to churn

# 2.0 Data Understanding

In [ ]:
# from pyforest import *

In [ ]:
import pandas as pd

In [ ]:
df = pd.read_csv('./bigml_59c28831336c6604c800002a.csv', index_col=0)
display(df.head())

account length  area code phone number international plan  \
state                                                              
KS                128        415     382-4657                 no   
OH                107        415     371-7191                 no   
NJ                137        415     358-1921                 no   
OH                 84        408     375-9999                yes   
OK                 75        415     330-6626                yes   

      voice mail plan  number vmail messages  total day minutes  \
state                                                             
KS                yes                     25              265.1   
OH                yes                     26              161.6   
NJ                 no                      0              243.4   
OH                 no                      0              299.4   
OK                 no                      0              166.7   

       total day calls  total day charge  total eve minutes  total eve calls  \
state                                                                          
KS                 110             45.07              197.4               99   
OH                 123             27.47              195.5              103   
NJ                 114             41.38              121.2              110   
OH                  71             50.90               61.9               88   
OK                 113             28.34              148.3              122   

       total eve charge  total night minutes  total night calls  \
state                                                             
KS                16.78                244.7                 91   
OH                16.62                254.4                103   
NJ                10.30                162.6                104   
OH                 5.26                196.9                 89   
OK                12.61                186.9                121   

       total night charge  total intl minutes  total intl calls  \
state                                                             
KS                  11.01                10.0                 3   
OH                  11.45                13.7                 3   
NJ                   7.32                12.2                 5   
OH                   8.86                 6.6                 7   
OK                   8.41                10.1                 3   

       total intl charge  customer service calls  churn  
state                                                    
KS                  2.70                       1  False  
OH                  3.70                       1  False  
NJ                  3.29                       0  False  
OH                  1.78                       2  False  
OK                  2.73                       3  False

In [ ]:
def analyze_dataset(df):
    # Dataset shape
    print("Shape of the dataset:", df.shape, '\n')
    print("========================================", '\n')

    # Missing values
    print("Null Values count:", df.isnull().sum(), '\n')
    print("========================================", '\n')

    # Duplicate values
    print("Number of duplicates:", len(df.loc[df.duplicated()]), '\n')
    print("========================================", '\n')

    # Target value count
    print("Count of each value of target:")
    print(df['churn'].value_counts(normalize=True), '\n')
    print("========================================", '\n')

    # Unique values
    print("The unique values per column are:")
    print(df.nunique(), '\n')
    print("========================================", '\n')

    # Dataset information
    print("Information about the dataset:")
    print(df.info())

    # distribution
    display(df.describe())


# Usage example
analyze_dataset(df)


Shape of the dataset: (3333, 20) 


Null Values count: account length            0
area code                 0
phone number              0
international plan        0
voice mail plan           0
number vmail messages     0
total day minutes         0
total day calls           0
total day charge          0
total eve minutes         0
total eve calls           0
total eve charge          0
total night minutes       0
total night calls         0
total night charge        0
total intl minutes        0
total intl calls          0
total intl charge         0
customer service calls    0
churn                     0
dtype: int64 


Number of duplicates: 0 


Count of each value of target:
False    0.855086
True     0.144914
Name: churn, dtype: float64 


The unique values per column are:
account length             212
area code                    3
phone number              3333
international plan           2
voice mail plan              2
number vmail messages       46
total day minutes       

account length    area code  number vmail messages  total day minutes  \
count     3333.000000  3333.000000            3333.000000        3333.000000   
mean       101.064806   437.182418               8.099010         179.775098   
std         39.822106    42.371290              13.688365          54.467389   
min          1.000000   408.000000               0.000000           0.000000   
25%         74.000000   408.000000               0.000000         143.700000   
50%        101.000000   415.000000               0.000000         179.400000   
75%        127.000000   510.000000              20.000000         216.400000   
max        243.000000   510.000000              51.000000         350.800000   

       total day calls  total day charge  total eve minutes  total eve calls  \
count      3333.000000       3333.000000        3333.000000      3333.000000   
mean        100.435644         30.562307         200.980348       100.114311   
std          20.069084          9.259435          50.713844        19.922625   
min           0.000000          0.000000           0.000000         0.000000   
25%          87.000000         24.430000         166.600000        87.000000   
50%         101.000000         30.500000         201.400000       100.000000   
75%         114.000000         36.790000         235.300000       114.000000   
max         165.000000         59.640000         363.700000       170.000000   

       total eve charge  total night minutes  total night calls  \
count       3333.000000          3333.000000        3333.000000   
mean          17.083540           200.872037         100.107711   
std            4.310668            50.573847          19.568609   
min            0.000000            23.200000          33.000000   
25%           14.160000           167.000000          87.000000   
50%           17.120000           201.200000         100.000000   
75%           20.000000           235.300000         113.000000   
max           30.910000           395.000000         175.000000   

       total night charge  total intl minutes  total intl calls  \
count         3333.000000         3333.000000       3333.000000   
mean             9.039325           10.237294          4.479448   
std              2.275873            2.791840          2.461214   
min              1.040000            0.000000          0.000000   
25%              7.520000            8.500000          3.000000   
50%              9.050000           10.300000          4.000000   
75%             10.590000           12.100000          6.000000   
max             17.770000           20.000000         20.000000   

       total intl charge  customer service calls  
count        3333.000000             3333.000000  
mean            2.764581                1.562856  
std             0.753773                1.315491  
min             0.000000                0.000000  
25%             2.300000                1.000000  
50%             2.780000                1.000000  
75%             3.270000                2.000000  
max             5.400000                9.000000

## Observations
* dataset contains 3333 rows and 20 columns

Variable Names and Descriptions : <br>
1. State: the US state in which the customer resides, indicated by a two-letter abbreviation
2. Account Length: the number of days that this account has been active
3. Area Code: the three-digit area code of the corresponding customer’s phone number
4. Phone: the remaining seven-digit phone number
5. Int’l Plan: whether the customer has an international calling plan: yes/no
6. VMail Plan: whether the customer has a voice mail feature: yes/no
7. VMail Message: presumably the average number

Our dataset has no Null values. Categorized of bla bla bla




# 3.0 Data Preparation

## 3.1 Data Cleaning

In [ ]:
# # Handle that class imbalance
# X = df.drop('churn', axis=1)
# y = df['churn'] # target

# X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)
# print(X_train.shape, y_train.shape)

In [ ]:
# from imblearn.over_sampling import SMOTE

# oversample = SMOTE()
# X_train_smote, y_train_smote = oversample.fit_resample(X_train, y_train)

# print(X_train_smote.shape, y_train_smote.shape)

In [ ]:
# y_train_smote.value_counts()

In [ ]:
# from lazypredict.Supervised import LazyClassifier
# from sklearn.model_selection import train_test_split

# X_train, X_test, y_train, y_test = train_test_split(X, y,test_size=.5,random_state =123)
# clf = LazyClassifier(verbose=0,ignore_warnings=True, custom_metric=None)
# models,predictions = clf.fit(X_train, X_test, y_train, y_test)
# models